docs/csharp/language-reference/compiler-messages/ref-modifiers-errors.md

```diff
@@ -0,0 +1,366 @@
+- **CS8338** - *The first '`in`' or '`ref readonly`' parameter of the extension method must be a concrete (non-generic) value type.*
+- **CS8977** - *Cannot use '`ref`', '`in`', or '`out`' in the signature of a method attributed with 'UnmanagedCallersOnly'.*
+- **CS9072** - *A deconstruction variable cannot be declared as a ref local*
+- **CS9199** - *A `ref readonly` parameter cannot have the Out attribute.*
+
+The following warnings indicate that a reference variable shouldn't be used, and might be unsafe:
+
+- **CS9184** - *'Inline arrays' language feature is not supported for inline array types with element field which is either a '`ref`' field, or has type that is not valid as a type argument.*
+- **CS9196** - *Reference kind modifier of parameter doesn't match the corresponding parameter in overridden or implemented member.*
+- **CS9197** - *Reference kind modifier of parameter doesn't match the corresponding parameter in hidden member.*
+- **CS9198** - *Reference kind modifier of parameter doesn't match the corresponding parameter in target.*
+- **CS9200** - *A default value is specified for `ref readonly` parameter, but `ref readonly` should be used only for references. Consider declaring the parameter as `in`.*
+- **CS9201** - *Ref field should be ref-assigned before use.*
+
+To fix the error, remove the reference variable where it isn't allowed:
+
+- Remove `in`, `ref` and `out` parameters from [indexers](../../programming-guide/indexers/index.md), [iterators](../../iterators.md), and [async methods](../../asynchronous-programming/index.md).
+- Remove [ref conditional expressions](../operators/conditional-operator.md#conditional-ref-expression) (`? :`) that include an [await](../operators/await.md).
+- Remove the `ref` modifier from the first parameter of a [extension method](../../programming-guide/classes-and-structs/extension-methods.md) where that type isn't a value type or a generic type constrained as a value type.
+- Either both or neither [conditional operator expressions] must be `ref` variables. Either remove `ref` from one expression, or add it to the other. If it's a `ref` conditional expression, both expressions must be the same type.
+- `ref` and `out` parameters can't have [default values](../../programming-guide/classes-and-structs/named-and-optional-arguments.md). Either remove the `ref` or `out` modifier, or remove the default value.
+- The range variable in a [LINQ query expression](../../linq/query-expression-basics.md) can't be passed by reference.
+- You can't deconstruct an object into reference variables. Replace the reference variables with value variables.
+- Implementing multiple interfaces where method overloads differ only on `ref` and `out`. For example, one interface declares `void M(ref int i)` and another interface declares `void M(out int i)`. A class can't implement both interfaces because the methods aren't distinguishable. You can only implement one of those interfaces.
```

Comment:
```suggestion
- You can't implement multiple interfaces where method overloads differ only on `ref` and `out`. For example, one interface declares `void M(ref int i)` and another interface declares `void M(out int i)`. A class can't implement both interfaces because the methods aren't distinguishable. You can only implement one of those interfaces.
```